### PR TITLE
[fenicsx render] - Add render method

### PR DESCRIPTION
**Pull request by Joao97ribeiro**
In this PR, I've introduced the rendering feature to the FEniCSx outputs. To achieve this, I've utilized the `create_2d_field_from_xdmf` method, which I've added to the `visualization` file.

I plan to create another PR to conduct testing for rendering in the NeuroStruct repository.

@baribeiro, please take a look at how I've implemented the XDMF file reading process. It might offer valuable insights for converting the XDMF file output into a PyG dataset

### CODE REVIEW

**Comment by Joao97ribeiro** [comment]
I'm adding a new method called `remove_time_values_from_xdmf` because the current version of PyVista lacks the ability to inspect time values in the reader (as mentioned in the [PyVista documentation](https://docs.pyvista.org/version/stable/api/readers/_autosummary/pyvista.XdmfReader.html)). In our specific use case, we only have one time instance, so by removing the lines with time values, we can ensure that it works correctly

File: inductiva/__init__.py
Comment (hpenedones) [inline inductiva/__init__.py:17]: I think that this had been commented to make sure there are no problems with imports of external libraries. Can you please check with @luispcunha and @IvanPombo that this won't break our "import inductiva"?

File: inductiva/structures/post_processing/fenicsx.py
Comment (hpenedones) [inline inductiva/structures/post_processing/fenicsx.py:8]: it is weird that you are inside the package inductiva but you are importing it, this seems like a circular import... No?

File: inductiva/structures/post_processing/fenicsx.py
Comment (hpenedones) [inline inductiva/structures/post_processing/fenicsx.py:6]: pyvista will have to go to an extra package dependency, right @IvanPombo @luispcunha ?

**Review by hpenedones — CHANGES_REQUESTED** [review]

File: inductiva/__init__.py
Comment (IvanPombo) [reply inductiva/__init__.py:17]: This is due to the gmsh not being anymore on the Client. I will check on this and you can notify me if it works on your side. We may need to follow the approach of this PR #538.

File: inductiva/structures/post_processing/fenicsx.py
Comment (IvanPombo) [reply inductiva/structures/post_processing/fenicsx.py:8]: I think it is the usual practice - Google Style Guide quote at 2.2 "Do not use relative names in imports. Even if the module is in the same package, use the full package name. This helps prevent unintentionally importing a package twice."

File: inductiva/utils/visualization.py
Comment (IvanPombo) [inline inductiva/utils/visualization.py:517]: With this created just go to the setup.cfg to add the structures_extra package with the extra pyvista.

File: inductiva/utils/visualization.py
Comment (IvanPombo) [inline inductiva/utils/visualization.py:564]: This should be an input variable.

File: inductiva/utils/visualization.py
Comment (IvanPombo) [inline inductiva/utils/visualization.py:567]: This as well.

File: inductiva/utils/visualization.py
Comment (IvanPombo) [inline inductiva/utils/visualization.py:554]: The cmap and show_edges should also be arguments. Provide this to the user, and you will be able to quickly change things as well :)

File: inductiva/structures/post_processing/fenicsx.py
Comment (luispcunha) [reply inductiva/structures/post_processing/fenicsx.py:8]: doing `from inductiva import z` where z is a submodule would also adhere to the style guide, but we've been getting feedback to do it like this, so this is how it looks like almost everywhere now: https://github.com/inductiva/inductiva/pull/439#discussion_r1286626468

File: inductiva/structures/post_processing/fenicsx.py
Comment (IvanPombo) [inline inductiva/structures/post_processing/fenicsx.py:22]: Isn't this function an utility function? It maybe useful for other future scenarios.

File: inductiva/structures/post_processing/fenicsx.py
Comment (IvanPombo) [inline inductiva/structures/post_processing/fenicsx.py:11]: The name of this class is too generic for FEniCSx, but actually your rendering is specific for the structures case. I think you won't be able to render all FEniCSx simulation scenarios with it, right?

File: inductiva/structures/post_processing/fenicsx.py
Comment (IvanPombo) [inline inductiva/structures/post_processing/fenicsx.py:78]: Imagine you have another rendering function or you call this function again, will you do this all the time? The information we are removing is it super important for the user to have, or can we just deliver the sim_output_path already without it?

File: inductiva/structures/post_processing/fenicsx.py
Comment (IvanPombo) [inline inductiva/structures/post_processing/fenicsx.py:89]: Can you highlight if the stress, strain and von_mises data is important for the user to manipulate and then render? Or they will just want to render and see the results?

File: inductiva/structures/post_processing/fenicsx.py
Comment (IvanPombo) [inline inductiva/structures/post_processing/fenicsx.py:100]: If I understand this, it will just create 21 graphs when field_names is None. Don't you think this is super cumbersome for the user (you)?

File: inductiva/__init__.py
Comment (Joao97ribeiro) [reply inductiva/__init__.py:17]: @hpenedones, I uncommented it because I thought that this problem had been resolved
@IvanPombo, thanks. I will now verify if everything is functioning correctly

File: inductiva/structures/post_processing/fenicsx.py
Comment (Joao97ribeiro) [reply inductiva/structures/post_processing/fenicsx.py:8]: @hpenedones, I'm following Luis's comment in the last PR (check out @luispcunha's comment with the link). Do you think it's better to approach it differently?

File: inductiva/structures/post_processing/fenicsx.py
Comment (Joao97ribeiro) [reply inductiva/structures/post_processing/fenicsx.py:6]: Do I need to include this in the `utils/optional_deps.py` file, right? I've already added: 

```
needs_structures_extra_deps = functools.partial(
    _needs_optional_deps,
    ["pyvista"],
    _missing_deps_msg("structures"),
)
```

Since I already have 'pyvista' in optional_deps, can I use it here?

File: inductiva/utils/visualization.py
Comment (Joao97ribeiro) [reply inductiva/utils/visualization.py:517]: like this, right?

```
needs_structures_extra_deps = functools.partial(
    _needs_optional_deps,
    ["pyvista"],
    _missing_deps_msg("structures"),
)
```

File: inductiva/structures/post_processing/fenicsx.py
Comment (Joao97ribeiro) [reply inductiva/structures/post_processing/fenicsx.py:22]: Yes, you're correct, I'll move it to the utils of the structure scenario, thanks!

File: inductiva/structures/post_processing/fenicsx.py
Comment (Joao97ribeiro) [reply inductiva/structures/post_processing/fenicsx.py:11]: If it's correct, it depends on what you're saving, whether it's displacements, stresses, strains, I'll renaming the class, thanks Ivan!

File: inductiva/structures/post_processing/fenicsx.py
Comment (Joao97ribeiro) [reply inductiva/structures/post_processing/fenicsx.py:100]: I understand your point, the idea here was to provide all the outputs as pictures in case people don't want to specify what they want to see. If you know what you want to see, you only need to specify the name. Do you think there's a better alternative?

File: inductiva/structures/post_processing/fenicsx.py
Comment (Joao97ribeiro) [reply inductiva/structures/post_processing/fenicsx.py:89]: It could be important, depending on what people want to study. Right now, our focus is primarily on rendering the outputs

File: inductiva/structures/post_processing/fenicsx.py
Comment (Joao97ribeiro) [reply inductiva/structures/post_processing/fenicsx.py:78]: Yes, it's not crucial. With this function, we're simply removing lines that contain the phrase 'time values,' so we're not losing any data. The only downside is having to repeat this process

File: inductiva/__init__.py
Comment (IvanPombo) [reply inductiva/__init__.py:17]: I believe it may be working for you, but not for other users. So I think we should try to pass the gmsh dependencies to the backend - specially because this is a scenario.

File: inductiva/structures/post_processing/fenicsx.py
Comment (IvanPombo) [reply inductiva/structures/post_processing/fenicsx.py:11]: But If I do another scenario and I provide the displacements, stresses and strains. Will I be able to use this?

File: inductiva/structures/post_processing/fenicsx.py
Comment (IvanPombo) [reply inductiva/structures/post_processing/fenicsx.py:91]: Then just remove it once (even in the backend if you want) and do not repeat it - it can cause some unneeded crashes.

File: inductiva/structures/post_processing/fenicsx.py
Comment (IvanPombo) [reply inductiva/structures/post_processing/fenicsx.py:113]: Yes, but with pyvista a little widget appears, so I need to just down 21 images before I move with my life. That is more painful, that selecting what I want to see. If you say to the user that he can see the stress, displacements, etc and the possible planes, then he can configure it himself. If you want to retrieve a list, that consider passing these args as a list.

File: inductiva/__init__.py
Comment (Joao97ribeiro) [reply inductiva/__init__.py:17]: Should I uncomment it now and then pass the Gmsh to the backend since this will be my next task?

Comment (IvanPombo) [reply inductiva/__init__.py:17]: Just uncomment, only if you are not using the gmsh anymore. Otherwise, and if you want to be able to merge keep it commented while you work. We will be deploying a new version of the package today and we do not want that to break it apart. Thanks :)

**Comment by Joao97ribeiro** [comment]
Thank you all for the suggestions; I've implemented all of them. I'm currently experiencing some pytest errors, and I believe they may be related to the Gmsh issue we discussed earlier, correct? Could you please provide your suggestions on how to resolve this bug? Thanks

**Comment by IvanPombo** [comment]
> Thank you all for the suggestions; I've implemented all of them. I'm currently experiencing some pytest errors, and I believe they may be related to the Gmsh issue we discussed earlier, correct? Could you please provide your suggestions on how to resolve this bug? Thanks

Remove the structures and the output_class = inductiva.structures....  this is crashing the pytest because the output_consts is being called from the package __init__, and then it checks every required package inside the structures module. Maybe adding the optional_deps can also insure that it passes the test when gmsh does not exist. But given that you will pass to the backend I would remove it from now.

**Comment by Joao97ribeiro** [comment]
Thanks, @IvanPombo! With your suggestion, the pytest issue has disappeared. @luispcunha, could you please check if the YAPF error is related to the api_client?

**Review by IvanPombo — COMMENTED** [review]
thanks for the updates @Joao97ribeiro! Everything is looking much more clean! I still have some notes that I would like for you to address in order to provide more readability. After them, it will be ready to merge!

File: inductiva/structures/post_processing/fenicsx_post_processing.py
Comment (Joao97ribeiro) [reply inductiva/structures/post_processing/fenicsx_post_processing.py:67]: Sorry, I already changed it to pv_multiblock. It is by pyvista.MultiBlock: https://docs.pyvista.org/version/stable/api/core/_autosummary/pyvista.MultiBlock.html#pyvista-multiblock

File: inductiva/utils/visualization.py
Comment (Joao97ribeiro) [reply inductiva/utils/visualization.py:519]: Good point! I've checked the `set_active_scalars function`, and it works with various types of [pyvista.DataSet](https://docs.pyvista.org/version/stable/api/core/_autosummary/pyvista.DataSet.html#pyvista.DataSet)

https://docs.pyvista.org/version/stable/api/core/_autosummary/pyvista.DataSet.set_active_scalars.html#pyvista-dataset-set-active-scalars

File: inductiva/structures/post_processing/fenicsx_post_processing.py
Comment (Joao97ribeiro) [reply inductiva/structures/post_processing/fenicsx_post_processing.py:63]: Why do you mention that it will be used and reused? People can provide the fields as a list, so this function is typically used only once, right?

File: inductiva/structures/post_processing/fenicsx_post_processing.py
Comment (Joao97ribeiro) [reply inductiva/structures/post_processing/fenicsx_post_processing.py:39]: I know, but it's the best way to avoid the pylint error: W0102: Dangerous default value [] as argument (dangerous-default-value). If we use ["von_mises"], we will encounter this error

**Comment by Joao97ribeiro** [comment]
Thanks again @IvanPombo for the suggestions! Please let me know if there are more things that I can improve before the merge

File: inductiva/structures/post_processing/fenicsx_post_processing.py
Comment (IvanPombo) [reply inductiva/structures/post_processing/fenicsx_post_processing.py:71]: Every time you call render you will re-do this again. But I can also want to render things in sequence.

File: inductiva/structures/post_processing/fenicsx_post_processing.py
Comment (IvanPombo) [inline inductiva/structures/post_processing/fenicsx_post_processing.py:84]: Just set this as a global variable...

**Review by IvanPombo — APPROVED** [review]
LGTM! Thanks for the good work!

File: inductiva/structures/post_processing/fenicsx_post_processing.py
Comment (Joao97ribeiro) [reply inductiva/structures/post_processing/fenicsx_post_processing.py:71]: I understand your point. With this suggestion, it's impossible to delete this file after rendering, right? In this way, we can delete this file, and I think it's important delete it because it's the same as with the results, but it doesn't have the time value information (which in this problem doesn't mean anything because it's just one instance). What do you think?

Comment (IvanPombo) [reply inductiva/structures/post_processing/fenicsx_post_processing.py:71]: But you could change the existing file instead of creating a copy. If you just lose one line, that is not important from the previous discussion, then do it on the initialization of the object.

File: inductiva/structures/post_processing/fenicsx_post_processing.py
Comment (Joao97ribeiro) [reply inductiva/structures/post_processing/fenicsx_post_processing.py]: The '0' represents the index of the array.  When we select a field, we have only one array -> index 0

**Review by Joao97ribeiro — COMMENTED** [review]
I will now proceed with the merge and focus on implementing mesh generation on the backend side

**Comment by Joao97ribeiro** [comment]
@hpenedones, I believe I require your approval to proceed with the merge. Whenever you have some time, please take a look

**Review by hpenedones — APPROVED** [review]
Thanks João!

**Pull request merged**
